### ISSUE TITLE
Custom invoice layout

### ISSUE THREAD

**Issue by nfq** [bytefury/crater#128]
Is there a way to custom the complete design and layout of the Estimates and invoices? Kind of important in my opinion. I won't be able to use the app in production yet, until I ca nestle the invoices exactly as I choose. 

Thanks for the great work so far!

**Comment by mohitpanjwani** [comment]
Sorry! As of now there’s no way to do that also,  this feature isn’t so easy to implement. 

We are looking to implement a white-labeling feature which allow the user to customise the colors and fonts etc but allowing customisations on layout like positioning and sizing isn’t so simple.

**Comment by nfq** [comment]
Thanks for the reply!

So, I haven't looked deeply through the code: but, when I found Crater, and saw it was web-based, I immediately thought HTML and CSS, awesome, I can style my invoices as I choose, unlike other bookkeeping systems with shitty template implementation.

Saying that. where does difficulty lie? In the PDF generation?

For me I can't use Crater yet, because I need to format my invoices in a certain way (I live and work in Switzerland).

Saying all of that, how can I help? I'd be more than happy to implement some code.

**Comment by mohitpanjwani** [comment]
Great, yeah all the invoice layouts are built with HTML/CSS but we use DOMPDF which doesn't support a lot of new CSS magic and sometimes you gotta use to the ancient table-based layout. 

Also, note that it involves a lot of trial and error in order to get it right :) 

You can customize the layout file according to your preference.

**Comment by nfq** [comment]
Ouch, Table layout. That's gotta hurt. So, where are the layout templates? Isn't obvious when browsing through the code.

**Comment by mohitpanjwani** [comment]
Sorry for the late reply. I was away on vacation :) 

You can find the invoice layouts here: `resources/views/app/pdf/invoice/invoice1.blade.php`

**Comment by HamzaDevz** [comment]
Yeah I think the best is to create a new layout like "invoice4.blade.php" and customize everything

**Comment by nfq** [comment]
I haven't used Crater in a while. Has the invoice design customisation, other than the suggestion above, been refined at all since the latest updates? I'm still on 2.1.0

**Comment by dkern** [comment]
Found this issue while searching for custom layouts for crater. Sad to read that this not an option currently. Therefore Crater is not usable for me. Everything else would be a perfect fit for my needs. But i will not send my invoices in this pinkish style ever! ;)

There should be some options to to changes in the future. I have no problem with dompdf an know that this has some downsides.

**Comment by nfq** [comment]
@dkern same for me. I mean, I am really glad this app exists and I think the developers have done a great job otherwise.

But having the ability to design the invoice would be necessary for some people I think, including me :)

**Comment by melistik** [comment]
Hi, maybe my little howto helps. I'm none main php-developer or contributor to this project but like it's feature set and look&feel. Same to you i needed to customize my invoice template so i've taken a look in the code and figured out how to get a new custom template in place:

The list of template options it stores in database:
- create a new entry in invoice_templates
  - `INSERT INTO invoice_templates (name, view, path, created_at, updated_at) VALUES ('Template 4', 'invoice4', '/assets/img/PDF/Template4.png', CURRENT_TIMESTAMP(), CURRENT_TIMESTAMP())`
- place a new template preview in `public/assets/img/PDF/Template4.png`
- copy one of the templates in `resources/views/app/pdf/invoice` name it invoice4.blade.php
- start customizing the php file to let it look how you like

That's all hope i helps someone...

**Comment by melistik** [comment]
I've crafted a custom template with repeatable headers + footers...

https://gist.github.com/melistik/0c94452e1a5f7c8e59e0be2369a22858

![preview](https://user-images.githubusercontent.com/1936030/94265505-f01e7200-ff38-11ea-852c-ceb19a9800f2.png)

**Comment by dkern** [comment]
If you don't want someone to see your personal/bank details, you should really edit your screenshots. :wink:

Well, making templates with DOMPDF isn't that hard. The problem i have is, that this all is kind of a _hack_ on `crater` and this might be having issues when updating in the future. Even if it works for now, there should just be an easy way to implement own Templates. Most simple way would be a Textbox somewhere to place the template and just do some string replaces by `crater`. Bot for now, i would not use it like this ...

**Comment by oliversum-ch** [comment]
Thanks a lot, @melistik!
Do you know how to set a customized template as a standard invoice template? I don't want to overwrite "Template 1", but use "My Company Template" as standard.
Maybe @mohitpanjwani can help?
Thank you very much!

**Comment by makakken** [comment]
has someone found a way to use a custom font? i want to use "Source Sans Pro" but when i load it with @font-face css i get errors....

`ErrorException
Undefined index: `

**Comment by flayks** [comment]
+1 on this one, considering the open source side of the tool that would absolutely great to have a custom template option, coded or not.

**Comment by mahamedsaid51** [comment]
Hi

does anyone know how to make template 3 as default invoice template ?

**Comment by mahamedsaid51** [comment]
> Hi, maybe my little howto helps. I'm none main php-developer or contributor of this project but I like it's feature set and it's look&feel. Same to you I needed to customize my invoice template so I've taken a short look into the code and figured out how to get a new custom template in place:
> 
> The list of template options it stores in database:
> 
> * create a new entry in invoice_templates
>   
>   * `INSERT INTO invoice_templates (name, view, path, created_at, updated_at) VALUES ('Template 4', 'invoice4', '/assets/img/PDF/Template4.png', CURRENT_TIMESTAMP(), CURRENT_TIMESTAMP())`
> * place a new template preview in `public/assets/img/PDF/Template4.png`
> * copy one of the templates in `resources/views/app/pdf/invoice` name it invoice4.blade.php
> * start customizing the php file to let it look how you like
> 
> That's all hope i helps someone...

Hi melistik

do you know how i make the template 3 as default invoice template

**Comment by NicoHood** [comment]
Hey guys, I came here via google and I am quite lost.

What I need is to add a german tax note (Kleinunternehmer) and payment details to every invoice. How would I do this best? Are there templates available somewhere? Something to get started would be nice, or maybe there is even an upstream footer customization feature?

**Comment by makakken** [comment]
you could probably add it to "Company Address"-Customization, like i did:

<img width="828" alt="Bildschirmfoto 2022-03-31 um 20 09 38" src="https://user-images.githubusercontent.com/10586572/161131428-6b2897d5-e66f-418f-bc3e-e79fc07c372f.png">

you can find this option under: settings/customization

**Comment by makakken** [comment]
@NicoHood does this help? or do you need furhter customization?

**Comment by NicoHood** [comment]
That is also a nice trick!

I've customized the template, and added:
```html
        @if ($invoice->getCustomFieldBySlug('CUSTOM_INVOICE_KLEINUNTERNEHMER')->default_answer)
            <div class="kleinunternehmer">
                Gemäß § 19 Abs. 1 UStG wird keine Umsatzsteuer ausgewiesen.
            </div>
        @endif
```

And the css:
```css
        /* -- Kleinunternehmer -- */

        .kleinunternehmer {
            font-size: 12px;
            color: #595959;
            margin-top: 15px;
            margin-left: 30px;
            width: 442px;
            text-align: left;
            page-break-inside: avoid;
        }
```

This option can now be checked for each invoice and be disabled later on when the business grows.

**Comment by derkolarsky** [comment]
Hi everyone,

I'm brand new to the crater invoicing system and I'm really excited. I chose the self-hosted version and now I'm trying to build my own invoice template. I'm getting through it quite well (I'm not a programmer, so it takes much more time…) but now I'm stuck.
Is there a way to insert single billing address variables? By now I just found {!! $billing_address !!} but I'd like to have the single variables, like **company name**, **contact name**, **addressline**, **zip** and **city**. But I have no clue what the correct names are…

I couldn't find any kind of listing that might help about the variables' names…
Are these variables somewhere to be found? That would make things much easier.

Thanks in advance
Ralf

**Comment by albertvisuals** [comment]
> I've crafted a custom template with repeatable headers + footers...
> 
> https://gist.github.com/melistik/0c94452e1a5f7c8e59e0be2369a22858
> 
> ![preview](https://user-images.githubusercontent.com/1936030/94266351-4e982000-ff3a-11ea-97a7-d6679e6cceb2.png)

Thanks a lot for your template. Sadly it now returns an error message in crater: 
ErrorException
Attempt to read property "company" on null (View: XXX/resources/views/app/pdf/invoice/FOTOHOF2.blade.php)

Any idea to get to working again?

**Comment by GunFood** [comment]
I've used the notes field and added the custom field into it in settings -> notes. Maybe an option for you.
